### PR TITLE
Add support for EFS Mount Points

### PR DESCRIPTION
**Pull request by serpentblade**
Support EFS volume mounts, including new EFS Access Point API

### CODE REVIEW

**Comment by serpentblade** [comment]
This should resolve issue #181

I'm successfully using this in a production Jenkins environment right now that utilizes an ECS cluster to run agents.  I have not tested with Fargate so anybody willing to assist would be appreciated.

**Comment by ltamaster** [comment]
Hi @serpentblade 
I am getting this error, any idea? 

(using the credentials from the IAM role of the ec2 instance) 
```
Sep 03, 2020 1:51:51 PM INFO com.cloudbees.jenkins.plugins.amazonecs.ECSTaskTemplate$EFSMountPointEntry$DescriptorImpl doFillFileSystemIdItems
Exception fetching file systems for credentials=, regionName=us-east-1
java.lang.NullPointerException
	at java.base/java.util.Comparator.lambda$comparing$77a9974f$1(Comparator.java:469)
	at java.base/java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
	at java.base/java.util.TimSort.sort(TimSort.java:220)
	at java.base/java.util.Arrays.sort(Arrays.java:1515)
	at java.base/java.util.ArrayList.sort(ArrayList.java:1750)
	at com.cloudbees.jenkins.plugins.amazonecs.ECSTaskTemplate$EFSMountPointEntry$DescriptorImpl.doFillFileSystemIdItems(ECSTaskTemplate.java:1008)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:396)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:408)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:212)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:145)
	at org.kohsuke.stapler.MetaClass$11.doDispatch(MetaClass.java:536)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:766)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:898)
	at org.kohsuke.stapler.MetaClass$4.doDispatch(MetaClass.java:281)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:766)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:898)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:694)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:240)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1631)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:154)
	at org.jenkinsci.plugins.ssegateway.Endpoint$SSEListenChannelFilter.doFilter(Endpoint.java:248)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.telemetry.impl.UserLanguages$AcceptLanguageFilter.doFilter(UserLanguages.java:129)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.security.ResourceDomainFilter.doFilter(ResourceDomainFilter.java:76)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at io.jenkins.blueocean.ResourceCacheControl.doFilter(ResourceCacheControl.java:134)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at io.jenkins.blueocean.auth.jwt.impl.JwtAuthenticationFilter.doFilter(JwtAuthenticationFilter.java:61)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at com.smartcodeltd.jenkinsci.plugin.assetbundler.filters.LessCSS.doFilter(LessCSS.java:47)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:153)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:51)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at jenkins.security.SuspiciousRequestFilter.doFilter(SuspiciousRequestFilter.java:36)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:549)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:578)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1369)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:489)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1284)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:501)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:272)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at 
```

**Comment by ltamaster** [comment]
for me this is falling: ` allFileSystems.sort(Comparator.comparing(FileSystemDescription::getName));`

**Comment by serpentblade** [comment]
Hi @ltamaster,

I think the problem was my assumption that FileSystemDescription::getName would always return a String.  I think having no name and returning NULL instead is the source of this NPE.

Would you mind changing that line you found and testing for me?
```java
allFileSystems.sort(Comparator.comparing(FileSystemDescription::getName, Comparator.nullsFirst(Comparator.naturalOrder())));
``` 

That should hopefully provide a null-safe comparator that won't throw the NPE.

There's another similar sort line inside of `doFillAccessPointIdItems()` that should probably be updated in the same way:

```java
accessPoints.sort(Comparator.comparing(AccessPointDescription::getName, Comparator.nullsFirst(Comparator.naturalOrder())));
```

I hope that should get things working for you, and I'll push a commit with those fixes if you're able to confirm.

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java [hunk 1]
```diff
@@ -752,24 +777,69 @@ public ECSTaskTemplate merge(ECSTaskTemplate parent) {
...
+
+                EFSAuthorizationConfig efsAuthorizationConfig = null;
+                if (StringUtils.isNotEmpty(mount.accessPointId)) {
+                    efsAuthorizationConfig = new EFSAuthorizationConfig()
+                            .withAccessPointId(mount.accessPointId)
+                            .withIam(BooleanUtils.isTrue(mount.iam)
+                                    ? EFSAuthorizationConfigIAM.ENABLED
+                                    : EFSAuthorizationConfigIAM.DISABLED
+                            );
+                }
```
Comment (Chris9224) [inline src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java:795]: Perhaps to extract this as a method called getEFSAuthorizationConfig?

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java [hunk 2]
```diff
@@ -752,24 +777,69 @@ public ECSTaskTemplate merge(ECSTaskTemplate parent) {
...
+
+                EFSVolumeConfiguration efsVolumeConfiguration = new EFSVolumeConfiguration()
+                        .withFileSystemId(mount.fileSystemId)
+                        .withRootDirectory(mount.rootDirectory)
+                        .withAuthorizationConfig(efsAuthorizationConfig)
+                        .withTransitEncryption(BooleanUtils.isTrue(mount.transitEncryption)
+                                ? EFSTransitEncryption.ENABLED
+                                : EFSTransitEncryption.DISABLED
+                        );
```
Comment (Chris9224) [inline src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java:804]: We can also extract this as a method called getEFSVolumeConfiguration?
This is to improve readability.

**Comment by FnTm** [comment]
Hi @serpentblade , I don't want to rush you, but I did want to send good wishes and encouragements regarding this PR as I'd reeeeally like to use this in our Setup 🙂

**Comment by serpentblade** [comment]
Hi @FnTm, unfortunately I don't have much control over merging this work.  It doesn't sound like the current maintainers are around to approve and merge things.  If you'd like to use this new feature, you can pull down my branch, run `mvn package`, and manually upload the plugin to Jenkins.  You can find the plugin package generated in `target/amazon-ecs.hpi`.  This is how I have it running in my environments while waiting for it to be officially merged and released.

**Comment by FnTm** [comment]
@serpentblade Yes, it does look like they have their hands full at this time. I tried building your code as you described, but I'm getting this error. Does it look like it could something to do with your code changes?

```
[WARNING] Tests run: 3, Failures: 0, Errors: 0, Skipped: 3, Time elapsed: 0.001 s - in com.cloudbees.jenkins.plugins.amazonecs.ECSServiceTest
[INFO] Running com.cloudbees.jenkins.plugins.amazonecs.ECSSlaveTest
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 2.648 s <<< FAILURE! - in com.cloudbees.jenkins.plugins.amazonecs.ECSSlaveTest
[ERROR] com.cloudbees.jenkins.plugins.amazonecs.ECSSlaveTest.terminateRunningTask  Time elapsed: 1.289 s  <<< ERROR!
org.mockito.exceptions.misusing.MissingMethodInvocationException:

when() requires an argument which has to be 'a method call on a mock'.
For example:
    when(mock.getArticles()).thenReturn(articles);

Also, this error might show up because:
1. you stub either of: final/private/equals()/hashCode() methods.
   Those methods *cannot* be stubbed/verified.
   Mocking methods declared on non-public parent classes is not supported.
2. inside when() you don't call method on mock but on some other object.

	at com.cloudbees.jenkins.plugins.amazonecs.ECSSlaveTest.terminateRunningTask(ECSSlaveTest.java:31)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:598)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:832)

[ERROR] com.cloudbees.jenkins.plugins.amazonecs.ECSSlaveTest.terminate_ThrowsException_ignoreException  Time elapsed: 1.355 s  <<< ERROR!
org.mockito.exceptions.misusing.MissingMethodInvocationException:

when() requires an argument which has to be 'a method call on a mock'.
For example:
    when(mock.getArticles()).thenReturn(articles);

Also, this error might show up because:
1. you stub either of: final/private/equals()/hashCode() methods.
   Those methods *cannot* be stubbed/verified.
   Mocking methods declared on non-public parent classes is not supported.
2. inside when() you don't call method on mock but on some other object.

	at com.cloudbees.jenkins.plugins.amazonecs.ECSSlaveTest.terminate_ThrowsException_ignoreException(ECSSlaveTest.java:57)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:598)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:832)

[INFO] Running com.cloudbees.jenkins.plugins.amazonecs.ECSTaskTemplateTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in com.cloudbees.jenkins.plugins.amazonecs.ECSTaskTemplateTest
[INFO] Running com.cloudbees.jenkins.plugins.amazonecs.pipeline.ECSTaskTemplateStepExecutionTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.033 s - in com.cloudbees.jenkins.plugins.amazonecs.pipeline.ECSTaskTemplateStepExecutionTest
[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   ECSCloudTest.addDynamicTemplateRegistersTemplate:96 » NullPointer
[ERROR]   ECSCloudTest.removeJunkTemplateProducesNoError:117 » NullPointer
[ERROR]   ECSSlaveTest.terminateRunningTask:31 MissingMethodInvocation
when() requires ...
[ERROR]   ECSSlaveTest.terminate_ThrowsException_ignoreException:57 MissingMethodInvocation
[INFO]
[ERROR] Tests run: 27, Failures: 0, Errors: 4, Skipped: 3
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  34.349 s
[INFO] Finished at: 2020-09-21T10:15:11+03:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (default-test) on project amazon-ecs: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/janis/dev/amazon-ecs-plugin-serpentblade/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

**Comment by serpentblade** [comment]
@FnTm No, that doesn't appear to have to do with my changes.  Seems there's possibly a conflict with the Mockito version and whatever version of JDK you are running.  Rather than try to debug all of that, you can build by skipping tests instead:
`mvn package -DskipTests`

Here's my version of JDK for comparison:
```
[INFO] --- maven-enforcer-plugin:3.0.0-M3:display-info (display-info) @ amazon-ecs ---
[INFO] Maven Version: 3.6.2
[INFO] JDK Version: 1.8.0_222 normalized as: 1.8.0-222
[INFO] OS Info: Arch: x86_64 Family: mac Name: mac os x Version: 10.15.6
```

**Comment by FnTm** [comment]
@serpentblade Thank you for that tip! I'm not super familiar with `mvn` so that tip helped. And even better:
**Your PR worked perfectly! 🎉**
The only minor thing was that I had to set the Fargate task version to 1.4.0. I remember reading about it before, so it wasn't that hard for me to figure out why my Agent wasn't starting, but maybe a good idea to add that to the docs as well.

**Comment by FnTm** [comment]
Dangit, missed commenting here to prevent closing. This PR shouldn't be closed!

**Comment by FnTm** [comment]
@gokhanoner I noticed that you may be a maintainer of this repo. Could you maybe also take a look at this PR?

**Comment by FnTm** [comment]
@markyjackson-taulia I noticed that you merged a PR in this repo recently. Could you maybe also look into this PR? I missed the stale bot, but I think this shouldn't be closed, and should indeed go into production.

**Pull request closed without merging**